keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+
+#### Application Status
+
+1. For Applications that provide a [health check](#health-checks), UIs may use the `Status.Ready` condition to 
+surface the readiness of the application to users. If an Application has no health check, the `Status.Ready` field 
+will always be `UNKNOWN`.
+1. UIs may surface the installation status of an Application's components and dependencies by examining the set of 
+installed components in the Application's `Status.Installed` field.
+
+### Risks and Mitigations
+
+1. In order for the Application API to be useful, we must have broad adoption. In order to mitigate the risk of 
+a lack of adoption, we will gather consensus from developers in SIG Apps prior to promoting the API to Beta.
+1. As discussed in [Drawbacks](#drawbacks), our approach to the implementation of the Application Controller may 
+increase the CPU and memory footprint of the Controller Manager process. We will mitigate this risk by mirroring the 
+efficient implementation of the Garbage Collector.
+
+## Graduation Criteria
+
+1. The graduation criteria from Alpha to Beta is simply the consensus of opinion among SIG Apps and SIG CLI that the 
+feature is ready to be consumed by tool and UI developers. We expect developers to prototype against the API at alpha 
+level stability, but, as it will be disabled by default, we do not expect to be able to gather feedback from end users.
+2. The graduation criteria from Beta to GA is adoption and interoperability. Before graduating the feature to GA we 
+would like to see broad adoption by tools (e.g. Helm, Bitnami installer) and UIs (Kubernetes OSS UI, GKE, OpenShift, 
+Azure).
+
+## Implementation History
+
+### Planned
+
+- Kubernetes 1.10 - Initial implementation of the Application Kind and Application Controller in the apps/v1alpha 
+Group Version.
+- Kubernetes 1.11 - Promotion of the Application Kind and Application Controller to the apps/v1beta1 Group Version. The 
+Application Controller and its associated API are enabled by default. The Application Kind is deprecated in the 
+apps/v1alpha Group Version.
+- Kubernetes 1.12 - The Application Kind is removed from the apps/v1aplha1 Group Version.
+- Kubernetes 1.13 - The Application Kind is promoted to the apps/v1 Group Version and deprecated in the apps/v1beta1 
+Group Version.
+- Kubernetes 1.14 - Storage migration is complete, on upgrade, to the apps/v1 Group Version of the Application Kind and 
+this version is disabled by default.
+- Kuberentes 1.15 - The apps/v1beta1 version of the Application Kind is removed from the API.
+
+## Drawbacks
+
+### Increased Memory, CPU, and Network Requirements
+The Application Controller, as designed, like the Garbage Collector, must watch a large fraction of the API objects 
+stored on the API server. We must be careful to implement the Application Controller efficiently to minimize 
+the increased memory and CPU requirements of the core Controller Manager and to minimize pressure on the API server. 
```

Comment:
There are a few very, very large Helm charts -- most notably, the ones that install and manage OpenStack. These may have hundreds of individual components. You might use that as a litmus test for this sort of thing. I imagine that the Kolla Kubernetes chart, for example, would spawn a large number of watches.